src/shaders/doubleSphereCamera.frag

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// ------------ uniform ----------------------
+uniform highp vec2 FocalLength;
+uniform highp vec2 PrincipalPointOffset;
+uniform highp float Alpha;
+uniform highp float Xi;
+
+#if defined(COLOR_TEXTURE)
+uniform samplerCube ColorTexture;
+#endif
+
+#if defined(DEPTH_TEXTURE)
+uniform samplerCube DepthTexture;
+#endif
+
+// ------------ output -----------------------
+#if defined(COLOR_TEXTURE)
+layout(location = OUTPUT_ATTRIBUTE_LOCATION_COLOR) out vec4 fragmentColor;
+#endif
+
+void main(void) {
+  vec3 m;
+  m.xy = (gl_FragCoord.xy - PrincipalPointOffset) / FocalLength;
+  // MUST flip the x axis
+  m.x = -m.x;
+  float r2 = dot(m.xy, m.xy);
+  float sq1 = 1.0 - (2 * Alpha - 1.0) * r2;
+  if (sq1 < 0.0) {
+    discard;
+  }
+  m.z = (1 - Alpha * Alpha * r2) / (Alpha * sqrt(sq1) + 1.0 - Alpha);
+  float mz2 = m.z * m.z;
+  float sq2 = mz2 + (1 - Xi * Xi) * r2;
+  if (sq2 < 0.0) {
+    discard;
+  }
+
+  // unproject to get the ray direction
+  vec3 ray = (m.z * Xi + sqrt(sq2)) / (mz2 + r2) * m - vec3(0.0, 0.0, Xi);
+
+#if defined(COLOR_TEXTURE)
+  fragmentColor = texture(ColorTexture, normalize(ray));
+#endif
+#if defined(DEPTH_TEXTURE)
+  gl_FragDepth = texture(DepthTexture, normalize(ray)).r;
```

Comment:
Isn't there a depth conversion that needs to happen? The depth in the cubemap texture is the distance along the given cubemap side axis (right?). Shouldn't the depth rendered for this sensor be distance along the camera look axis?

Can you attach some depth renders converted to grayscale? Especially looking in the direction of cubemap corners.

Reply:
> Isn't there a depth conversion that needs to happen? The depth in the cubemap texture is the distance along the given cubemap side axis (right?). Shouldn't the depth rendered for this sensor be distance along the camera look axis?

The depth is a scale, not a vector. The value is like the value in the ordinary depth buffer in an ordinary image (see [here](https://github.com/facebookresearch/habitat-sim/blob/08f5dcae8cb9e94e0985f8134affdb0e8ce1cc2c/src/esp/gfx/CubeMap.cpp#L178))
(A good tutorial is here: https://learnopengl.com/Advanced-OpenGL/Depth-testing)

>Can you attach some depth renders converted to grayscale?

This will depend on this [diff](https://github.com/facebookresearch/habitat-sim/pull/1066), which needs big changes. But I will see what I can do.

>Especially looking in the direction of cubemap corners.

The cubemap is fixed in the camera space, NOT in the world space. You can imagine camera is locating in the center of this cube, and always looking at the -Z direction (Y is up, X is right). If camera rotates (in the world space), the cubemap rotates. In other words, cubemap needs to be rendered in every frame when camera is looking around.

Reply:
>The cubemap is fixed in the camera space, NOT in the world space. 

Ok, makes sense. My concern remains. Let's see a depth render with a wide enough FOV such that some cubemap edges/corners are "in frame".

Reply:
Yes, usually there are 8 edges (LT, LF, LB, FT, FB, RT, RF, RB) in the rendering frame. (L = left, R = Right, T = Top, B = bottom) It happens for example in the demo shown in the meeting.